perf(codegen): faster writing indentation

overlookmotel opened the pull request:
Write indentation faster. Previously was writing indentation to buffer with a call to `memset`, which is relatively expensive. Where indentation `<= 16` (common case), write 16 tabs with a single 16-byte XMM write, which is faster.

overlookmotel commented:
* **#7820** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/7820?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/7820?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=7820).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel commented:
+1% perf gain.

Boshen commented on crates/oxc_codegen/src/code_buffer.rs:
```diff
@@ -365,6 +374,55 @@ impl CodeBuffer {
+        /// 16 is largest register size (XMM) available on all x86_84 targets.
+        const CHUNK_SIZE: usize = 16;
+
+        #[cold]
+        #[inline(never)]
+        fn write_slow(code_buffer: &mut CodeBuffer, n: usize) {
+            code_buffer.buf.extend(std::iter::repeat(b'\t').take(n));
+        }
```
Prevent filling of the CPU instructions cache.

overlookmotel replied:
The idea is that 99% of the time, both these conditions are true:

1. The number of tabs to be inserted for indentation is `<= 16` (nesting functions/blocks/etc more than 16 deep in JS code is uncommon).
2. There is spare capacity in the buffer for at least 16 bytes (the buffer grows by doubling its capacity, so there's usually loads of capacity).

So we want to make the path for this common circumstance as fast as possible, and also as small as possible (in number of instructions) so `print_indent` can be inlined into the caller.

There's 2 things going on here:

`#[cold]` hints to compiler that `write_slow` is rarely called. This tells compiler to generate assembly for the branches on these 2 conditions where the common case is encoded as "do nothing, continue on to next instruction" - the fastest path. See the assembly for `print_tabs_v2` here: https://godbolt.org/z/e1EP5cnPc (scroll down)

`#[inline(never)]` is there for reason Boshen said. It tells compiler to store the code for `write_slow` far away from the rest of the code. `write_slow` will very rarely be called, so you don't want it loaded eagerly into CPU instruction cache. Space in instruction cache is better used by code which likely *will* be run soon.

To be honest, I don't know how much difference `#[inline(never)]` makes, or whether it's necessary. `#[cold]` is the more important one.

Unstable Rust has `likely` and `unlikely` intrinsics which allow writing this code in a simpler way. `likely(...)` tells compiler that the expression it's passed is likely to be true:

```rs
if likely(n <= CHUNK_SIZE && spare_capacity >= CHUNK_SIZE) {
    // Compiler assume this branch will be taken (fast path)
} else {
    // Assume this branch will not be taken (slow path)
}
```

But `likely` and `unlikely` currently don't work on stable Rust, so only way to achieve this pattern at present is the `#[cold]` attribute:

```rs
if n <= CHUNK_SIZE && spare_capacity >= CHUNK_SIZE {
    // Fast path for common case
} else {
    #[cold]
    fn slow() {
        // Slow path for uncommon case
    }
    slow();
}
```

See: https://users.rust-lang.org/t/is-cold-the-only-reliable-way-to-hint-to-branch-predictor/106509

I am no expert! I am still feeling my way on this stuff.

It also doesn't help that the CPU simulation which CodSpeed uses for benchmarking is pretty basic, and doesn't have a branch predictor. So the effect of the branch predictor optimization in this PR will not be showing in bench results for this PR. Probably the real world impact is greater than the +1% that CodSpeed shows.

overlookmotel replied:
PS This kind of micro-optimization usually isn't worth it. But in codegen a lot of the work it does is just pushing bytes to the output buffer. So "push to buffer" is a very hot path, called over and over, and so it becomes worthwhile optimizing.

Boshen commented:
### Merge activity

* **Dec 13, 12:16 AM EST**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.
* **Dec 13, 12:16 AM EST**: A user added this pull request to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc).

overlookmotel committed:
perf(codegen): faster writing indentation (#7820)

Write indentation faster. Previously was writing indentation to buffer with a call to `memset`, which is relatively expensive. Where indentation `<= 16` (common case), write 16 tabs with a single 16-byte XMM write, which is faster.
